feat: SWE Agent Implementation

JayQuimby opened the pull request:
## WIP: Implementing an agent that works similar to [SWE-Agent](https://github.com/princeton-nlp/SWE-agent)

### Agent Features:
- [x] Application Computer Interface - simplifies the commands required by the llm.
  - [x] Implement: 'read', 'write', 'browse', 'exit', 'edit', 'goto', 'scroll'
- [x] Think-Act prompting structure - guides the model to think about what its next step should be and then make code to do it
- [x] Short Term Memory - Allows the model to see it's last n steps before taking the next one
- [x] Tells the model about its current working dir, file, and line

### Help / Suggestions:
- Open to other ideas people might have on how to improve this
- This agent should be similar to SWE agent, so I would appreciate ideas staying related to that
- Any help or additional code would be much appreciated

### Links/context:
- [SWE-Agent Demo](https://www.youtube.com/watch?v=RJ6NN8Y-xok&ab_channel=OfirPress)
- [Repo](https://github.com/princeton-nlp/SWE-agent)
- Issue #570 

| Agent/model | SWE-bench % resolved |
| :---------------- | :------: |
| SWE-Agent (GPT-4) | 12.29% |
| Devin (25% of eval set) | 13.84% |
| Claude 3 Opus (RAG) | 3.79% |
| GPT-4 (RAG) | 1.44% |

Since SWE-Agent is so close to Devin in performance, I figure trying to emulate it and then iterate on it is the best route.

JayQuimby commented:
> Can you fix Github Issues whith this and create a PR?

What github issues are you talking about? can you provide a screenshot or something for me to go off of?

PierrunoYT commented:
> @PierrunoYT
> 
> > Can you fix Github Issues whith this and create a PR?
> 
> What github issues are you talking about? can you provide a screenshot or something for me to go off of?

I mean like the SWE-Agent which can fix Github Issues.

JayQuimby commented:
Ohhh, lol completely misread this the first time.

I do not think it is currently capable of doing that, not sure tho. I am just implementing the agent framework, but there will definitely need to be more work done on getting web pages, formatting, submitting pull requests etc. for this to do what you were talking about before...

PierrunoYT commented:
> Ohhh, lol completely misread that the first time.
> 
> I do not think it is currently capable of doing that, not sure tho. I am just implementing the agent framework, but there will definitely need to be more work done on getting web pages, formatting, submitting pull requests etc. for this to do what you were talking about before...

Oh okay. Waiting for it :)

foragerr commented:
@PierrunoYT There's some discussion and work [here about using OpenDevin to work on OpenDevin GitHub issues](https://github.com/OpenDevin/OpenDevin/issues/198#issuecomment-2040890496). The agent quality with the currently implemented agents is pretty poor though.

I'm excited to see a variant of SWE-agent implemented.

rbren commented:
Rad! @JayQuimby I'm curious how well it seems to work in practice. Have you thrown any examples at it?

rbren commented on agenthub/thinkact_agent/plans.md:
```diff
@@ -0,0 +1,36 @@
+    'description': 'This will allow you to modify files within your working directory. Usage: "edit example.txt" -> opens the example.txt file and shows the first 100 lines',
+},
+
+'goto': {
+    'params': '<line_num>',
+    'description': 'This will allow you to go through a file to any line. Usage: "Goto 124" -> returns lines 124-224 within current file',
+},
+
+'scroll_up': {
+    'params': '',
+    'description': 'When you are in a file you can see the 100 lines above your current view. Usage: "scroll_up" -> returns the 100 lines above what you were reading',
+},
+
+'scroll_down': {
+    'params': '',
+    'description': 'When you are in a file you can see the 100 lines below your current view. Usage: "scroll_down" -> returns the 100 lines below what you were reading',
+},
```
We should translate our current `read` command here, e.g. adding `start_line` and `end_line` parameters

`scroll_up` and `scroll_down` seem superfluous--the agent could just `goto` cur+100. Curious why SWE-agent did this. IMO we should handle the `scroll` actions inside the Agent python rather than abstracting them out into new Actions

JayQuimby replied:
I was watching a deep dive with one of the devs and he was saying that the scroll ability helps a lot when the model does not know/ recall the line it needs to go to. for example, they showed it searching for a function after it found that function in a file by scrolling. I do agree that it can probably be better achieved with 'goto' just thought I would add it to stay in line with the original swe-agent.

0xdevalias replied:
A few arbitrary things I've been reading lately seemed to imply that the models often have issues with specific lines numbers/etc; so I suspect they opted for this method for a similar reason.

rbren replied:
Yeah agree we should stick closely to their implementation--they're doing something right!

rbren replied:
I don't see any implementation of the `scroll` commands. I would expect our `read` command to be modified to accept a position and number of lines (or startline/endline)

rbren commented on agenthub/thinkact_agent/plans.md:
```diff
@@ -0,0 +1,36 @@
+    'description': 'When you are in a file you can see the 100 lines below your current view. Usage: "scroll_down" -> returns the 100 lines below what you were reading',
+},
+
+'modify': {
+    'params': '<start_line>:<end_line> "<replacement>"',
+    'description': 'This will make changes to a file by deleting all lines from <start_line> to <end_line> and replacing them with <replacement>',
+}
```
We should change our `write` command to support start and end lines

JayQuimby commented:
> Rad! @JayQuimby I'm curious how well it seems to work in practice. Have you thrown any examples at it?

@rbren, I have not tested it with gpt-4 / claude yet because it still needs some work. I will keep working on it, hope to have something usable in the next few days.

rbren commented on agenthub/thinkact_agent/plans.md:
````diff
@@ -0,0 +1,36 @@
+
+## 1. Add in all of the following commands to ACI
+```
+'search_for': {
+    'params': '<keywords>',
+    'description': 'Will allow you to search your working directory for files and folders that match your <keyword>.',
+}
````
Definitely want to add a `search_for` command. IMO it should be called `search_code` so the LLM doesn't confuse it with searching on Google

foragerr commented:
Just want to drop this for passers-by, totally worth a watch if you have 20 mins: https://www.youtube.com/watch?v=RJ6NN8Y-xok

JayQuimby commented:
@rbren can you look at this again, I think we are getting close to something that is usable. There are still some bugs we are working on fixing but the overall agent is there.

rbren commented:
Exciting! Will take a look

rbren commented on agenthub/thinkact_agent/__init__.py:
```diff
@@ -0,0 +1,4 @@
+from opendevin.agent import Agent
+from .agent import ThinkActAgent
+
+Agent.register('ThinkActAgent', ThinkActAgent)
```
Can we call it SWEAgent, so folks know where it came from?

rbren commented on opendevin/action/fileop.py:
```diff
@@ -1,49 +1,153 @@
+PATH_PREFIX = '/workspace/'
+
+
+def validate_file_content(file_path: str, content: str) -> str:
```
Do we need this given that it's unimplemented?

Maybe we wait for https://github.com/OpenDevin/OpenDevin/issues/1043

JayQuimby commented:
@rbren 
1. I have tested this with gpt-4 a few times and it seems to be working well, local models do not appear to work well no matter what I try to do. 
2. The custom bash commands do not appear to be working (search_file, search_dir, find_file...). I think I might just implement them in python instead of bash, curious on your thoughts there. 
3. I renamed everything like you said and goto, scroll commands are implemented. I think the prompt ordering/phrasing might need some work but it is quite functional as is. 
4. All the autogenerated linter changes that were unnecessary are gone
5. I have some improvement ideas that I want to implement, but I wanted to get this finished as a starting point. Let me know if you think there are any things that need changing.

rbren commented on opendevin/action/fileop.py:
```diff
@@ -19,13 +24,28 @@ def resolve_path(file_path):
 class FileReadAction(ExecutableAction):
+    """
+    Reads a file from a given path up to 100 lines.
+    Default lines 0:100
```
Default here definitely needs to be the full file--some agents aren't aware of the start/end params

rbren commented on opendevin/action/fileop.py:
```diff
@@ -19,13 +24,28 @@ def resolve_path(file_path):
         path = resolve_path(self.path)
         with open(path, 'r', encoding='utf-8') as file:
-            return FileReadObservation(path=path, content=file.read())
+            all_lines = file.readlines()
+            total_lines = len(all_lines)
+            if total_lines >= 100:
+                end_index = self.start_index + 100 if total_lines - \
+                    self.start_index - 100 >= 0 else -1
+                code_slice = all_lines[self.start_index: end_index]
+            else:
+                code_slice = all_lines[:]
+            code_view = ''.join(code_slice)
```
maybe we pass in `total_lines` as a parameter?

JayQuimby replied:
I like this idea!

rbren commented:
@JayQuimby this is looking great!

Only concern is the changing behavior of FileReadAction

foragerr commented:
@JayQuimby Are you testing via the UI, or some other way?

JayQuimby commented:
@rbren 
> Only concern is the changing behavior of FileReadAction

Changed this so the default behavior is just reading the whole file, also added negative indices to the command detection so the model can specify from any line to the end of the file.

JayQuimby commented:
> @JayQuimby Are you testing via the UI, or some other way?

@foragerr I am testing with both gpt-4 and codellama using the sandbox. localhost:3001

rbren commented on opendevin/action/fileop.py:
```diff
@@ -19,13 +24,36 @@ def resolve_path(file_path):
+            else:
+                all_lines = file.readlines()
+                num_lines = len(all_lines)
+                begin = max(0, min(self.start, num_lines - 2))
```
why is this `-2`?

assertion commented:
Not sure whether I missed something? Does OpenDevin run the SWE-bench with the new SWE agent？ what's the resolved rate of OpenDevin compared with the other Devin like open-source projects.
